lib/srv/bpf_test.go

```diff
@@ -0,0 +1,871 @@
+				}
+			}
+
+			// Check that the emitted events have expected contents.
+			for _, event := range recordedEvents {
+				switch e := event.(type) {
+				case *apievents.SessionCommand:
+					t.Logf("  command event: Command=%s Path=%s Args=[%s] NArgs=%d", e.BPFMetadata.Program, e.Path, quoteStrings(e.Argv), len(e.Argv))
+
+					checkCommandEvent(t, e, programPaths, commandArgs)
+				case *apievents.SessionDisk:
+					t.Logf("  disk event: Command=%s Path=%s", e.BPFMetadata.Program, e.Path)
+
+					if checkDiskEvent(t, e, programLibs, true) {
+						continue
+					}
+					checkDiskEvent(t, e, commandOpens, false)
+				case *apievents.SessionNetwork:
+					t.Logf("  network event: Command=%s SrcAddr=%s DstAddr=%s DstPort=%d", e.BPFMetadata.Program, e.SrcAddr, e.DstAddr, e.DstPort)
+
+					checkNetworkEvent(t, e, commandDstAddrs)
+				}
+			}
+
+			// Check that expected events were found the expected number of times.
+			for cmd, count := range commandArgs {
+				if count > 0 {
+					t.Errorf("error: command event for %s was expected %d more times", cmd, count)
+				}
+			}
+			for cmd, paths := range commandOpens {
+				for _, path := range paths {
+					if path.count > 0 {
+						t.Errorf("error: disk event for program %q opening %q was expected %d more times", cmd, path.value, path.count)
+					}
+				}
+			}
+			for cmd, addrs := range commandDstAddrs {
+				for _, addr := range addrs {
+					if addr.count > 0 {
+						t.Errorf("error: network event for program %q with destination address %q was expected %d more times", cmd, addr.value, addr.count)
+					}
+				}
+			}
+			for cmd, libs := range programLibs {
+				for _, lib := range libs {
+					if lib.count > 0 {
+						t.Errorf("error: disk event for program %q opening library %q was expected %d more times", cmd, lib.value, lib.count)
+					}
+				}
+			}
+		})
+	}
+}
+
+type countedValue[T comparable] struct {
+	value T
+	count int
+}
+
+func makeCounted[T comparable](s []T, count int) []countedValue[T] {
+	countedValues := make([]countedValue[T], len(s))
+	for i, v := range s {
+		countedValues[i] = countedValue[T]{value: v, count: count}
+	}
+	return countedValues
+}
+
+// commandKey returns a string that can be used as a unique map key for
+// a command, given the command name and arguments. The returned string
+// describes the command clearly when printed.
+func commandKey(program string, args []string) string {
+	return fmt.Sprintf("%s [%s]", program, quoteStrings(args))
+}
+
+func newBPFService(t *testing.T) bpf.BPF {
+	t.Helper()
+
+	bpfSrv, err := bpf.New(&servicecfg.BPFConfig{
+		Enabled:    true,
+		CgroupPath: t.TempDir(),
+	})
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		const restarting = false
+		require.NoError(t, bpfSrv.Close(restarting))
+	})
+
+	return bpfSrv
+}
+
+func handleConnections(l net.Listener) {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) {
+				return
+			}
+		}
+		conn.Close()
```

Comment:
Don't we still need the continue for all other error types? Wouldn't we panic on the conn.Close call below otherwise?
```suggestion
		conn, err := l.Accept()
		if err != nil {
			if errors.Is(err, net.ErrClosed) {
				return
			}
			continue
		}
		conn.Close()
```